File: enero2025/ene24_C/src/Documento.cpp
#include "Documento.h"

int Documento::contador = 1; // ← Inicialización del static

Documento::Documento(char *t, char *a) {
    idDocumento = contador++;
    titulo = new char[strlen(t) + 1];
    strcpy(titulo, t);
    autor = new char[strlen(a) + 1];
    strcpy(autor, a);
}

Documento::Documento(const Documento& d) {
    idDocumento = contador++; // ← Nuevo ID
    titulo = new char[strlen(d.titulo) + 1];
    strcpy(titulo, d.titulo);
    autor = new char[strlen(d.autor) + 1];
    strcpy(autor, d.autor);
}

Documento::~Documento() {
    delete[] titulo;
    delete[] autor;
}

Documento& Documento::operator=(const Documento& d) {
    if (this != &d) {
        idDocumento = d.idDocumento; // ← Mismo ID en asignación
        delete[] titulo;
        delete[] autor;
        titulo = new char[strlen(d.titulo) + 1];
        strcpy(titulo, d.titulo);
        autor = new char[strlen(d.autor) + 1];
        strcpy(autor, d.autor);
    }
    return *this;
}

bool Documento::operator==(const Documento &d) const {
    return (idDocumento == d.idDocumento &&
            strcmp(titulo, d.titulo) == 0 &&
            strcmp(autor, d.autor) == 0);
}

ostream& operator<<(ostream &s, const Documento &d) {
    s << d.idDocumento << ": " << d.titulo << " (" << d.autor << ")";
    return s;
}

File: enero2025/ene24_java/src/ene24_java/Documento.java
package ene24_java;

public class Documento implements Cloneable {
    private static int contador = 1;
    private int idDocumento;
    private String titulo;
    private String autor;
    
    // Constructor principal
    public Documento(String t, String a) {
        idDocumento = contador;
        contador++;
        titulo = t;
        autor = a;
    }
    
    // Constructor de copia (para new Documento(d1))
    public Documento(Documento d) {
        idDocumento = contador;  // Nuevo ID
        contador++;
        titulo = d.titulo;
        autor = d.autor;
    }
    
    // Constructor protegido para clone (mismo ID)
    protected Documento(Documento d, int notUsed) {
        idDocumento = d.idDocumento;  // Mismo ID
        titulo = d.titulo;
        autor = d.autor;
    }
    
    @Override
    public Object clone() {
        return new Documento(this, 1);
    }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        
        Documento d = (Documento) o;
        return idDocumento == d.idDocumento && 
               titulo.equals(d.titulo) && 
               autor.equals(d.autor);
    }
    
    @Override
    public String toString() {
        return idDocumento + ": " + titulo + " (" + autor + ")";
    }
}